### DISCUSSION TITLE
Linux: btop fails to find AMD GPU

### DISCUSSION THREAD

**Discussion by bdols** [Homebrew/discussions#5793] [Linux] [answered]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8c868ba70e526f1f00561470d58fcbaddb1b4e48
Last commit: 2 days ago
Branch: stable
Core tap JSON: 08 Dec 00:18 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 24
SUDO_ASKPASS: /usr/libexec/openssh/gnome-ssh-askpass
Homebrew Ruby: 3.3.6 => /var/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 24-core 64-bit zen3
Clang: 19.1.5
Git: 2.47.0 => /bin/git
Curl: 8.6.0 => /bin/curl
Kernel: Linux 6.11.3-200.fc40.x86_64 x86_64 GNU/Linux
OS: Bluefin-dx 40 (FROM Fedora Silverblue) (Archaeopteryx)
Host glibc: 2.39
/usr/bin/gcc: 14.2.1
/usr/bin/ruby: N/A
glibc: 2.35_1
gcc@11: N/A
gcc: 14.2.0_1
xorg: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

for comparison, I've compiled btop from the v1.4.0 tag and also installed the btop formula. (to reproduce, delete any .config/btop/btop.conf, which will be regenerated on first run),  I have the rocm-smi RPM installed and the compiled btop appears to find /usr/lib64/librocm_smi64.so.1 correctly, and I did not need to clone https://github.com/rocm/rocm_smi_lib.git into the btop checkout as described in the project's README.

separately, I have another linux install, PopOS, with a NVIDIA GPU, and brew-provided btop finds the GPU fine (again I had to remove any existing .config/btop/btop.conf as it didn't find the GPU otherwise)

what's different about the brew provided btop?

**Comment by ZhongRuoyu** [comment, marked as answer]
This means that support for AMD GPU is not provided by Homebrew's btop. When there is no popular demand, it doesn't seem ideal to add the rocm-smi dependency to btop for everyone, so I suggest that you use your custom build if you need AMD GPU support.

Also, from [upstream's README](https://github.com/aristocratos/btop?tab=readme-ov-file#optional-dependencies-needed-for-gpu-monitoring-only-linux):

> - **AMD**
> If you have an AMD GPU `rocm_smi_lib` is required, which may or may not be packaged for your distribution.

**Comment by bdols** [comment]
I did not need to compile the rocm_smi_lib though, the code attempts to autodetect the library: https://github.com/aristocratos/btop/blob/main/src/linux/btop_collect.cpp#L1313

I run strace on both and this is what it's supposed to do, I think, finding the OS installed rocm_smi in /lib64:
out.1494275:openat(AT_FDCWD, "/opt/rocm/lib/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
out.1494275:openat(AT_FDCWD, "/lib64/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
out.1494275:openat(AT_FDCWD, "/usr/lib64/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
out.1494275:openat(AT_FDCWD, "/lib64/librocm_smi64.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
out.1494275:openat(AT_FDCWD, "/usr/lib64/librocm_smi64.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
out.1494275:openat(AT_FDCWD, "/lib64/librocm_smi64.so.1.0", O_RDONLY|O_CLOEXEC) = 3

while homebrew's btop doesn't search in /usr/lib64 or /lib64:
brew.out.1502005:openat(AT_FDCWD, "/opt/rocm/lib/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/lib/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/lib/librocm_smi64.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/librocm_smi64.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/lib/librocm_smi64.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/lib/librocm_smi64.so.5", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/librocm_smi64.so.1.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/lib/librocm_smi64.so.1.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/lib/librocm_smi64.so.1.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/librocm_smi64.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/lib/librocm_smi64.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.out.1502005:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/lib/librocm_smi64.so.6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)

`ldconfig -p | grep librocm` only shows:
        librocm_smi64.so.1 (libc6,x86-64) => /lib64/librocm_smi64.so.1

I see that the runtime search path is set as follows, which makes sense, but the strace shows that it's looking for  /home/linuxbrew/.linuxbrew/Cellar/**glibc/2.35_1/**lib/librocm_smi64.so.6 which are not in the rpath list (from `readelf -d $(which btop)`):
 0x000000000000000f (RPATH)              Library rpath: [/home/linuxbrew/.linuxbrew/Cellar/btop/1.4.0/lib:/home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current:/home/linuxbrew/.linuxbrew/lib]

meanwhile on the nvidia system, it finds the libnvidia-sm library:
brew.3934819:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/libnvidia-ml.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.3934819:openat(AT_FDCWD, "/home/linuxbrew/.linuxbrew/lib/libnvidia-ml.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
brew.3934819:openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libnvidia-ml.so", O_RDONLY|O_CLOEXEC) = 3
and % ldconfig -p | grep nvidia-ml shows:
        libnvidia-ml.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ml.so.1
        libnvidia-ml.so.1 (libc6) => /lib/i386-linux-gnu/libnvidia-ml.so.1
        libnvidia-ml.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libnvidia-ml.so
        libnvidia-ml.so (libc6) => /lib/i386-linux-gnu/libnvidia-ml.so

it seems that dlopen could find it if the rocm_lib package has an additional symlink for librocm_smi64.so to librocm_smi64.so.1 (because btop includes librocm_smi64.so as a library to search for), but again, my hand-compiled binary is able to find the library.
indeed, the man page for ldconfig suggests that the problem could be that the rocm_lib package will break compatibility unless it follows this pattern:
`
       ldconfig will look only at files that are named lib*.so* (for regular shared objects) or ld-*.so* (for the dynamic loader itself).  Other files will be ignored.  Also, ldconfig expects a certain pattern to how the symbolic links are set up, like this example, where the middle file (lib‐
       foo.so.1 here) is the SONAME for the library:

           libfoo.so -> libfoo.so.1 -> libfoo.so.1.12

       Failure to follow this pattern may result in compatibility issues after an upgrade.
`
